### PR TITLE
Implement basic bpf() wrappers for map creation and program loading

### PR DESCRIPTION
**Pull request by jscherpelz-juniper**
Add a new cc_library target to hold bpf wrappers. Currently has a
working test for map create/insert/update/lookup. An incomplete
program load wrapper is present but untested.

Also add basic posix syscall infrastructure.

### CODE REVIEW

File: lib/bpf/bpf.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+#include "lib/error/status_or.h"
+#include "lib/posix/syscall.h"
+
+namespace bpf {
+
+// bpf() syscall wrapper. No C standard library support currently exists.
+//
+// See "man 2 bpf" for details.
+inline error::StatusOr<int> Bpf(const int cmd, bpf_attr* const attr) {
```
Comment (epeer-juniper) [inline lib/bpf/bpf.h]: Should we provide a more specific type for cmd here? The return types are somewhat cleaner already, although I guess we're still exposing int instead of something more specific inside StatusOr. Or, do we wrap these calls in cleaner high level APIs later?

File: lib/bpf/map_create.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// BPF_MAP_CREATE wrapper.
+// Create a map of 'map_type' from keys of type KeyT to values of type ValueT
+// and maximum capacity of 'max_entries'.
+//
+// Return an owned file descriptor on success or an error status on failure.
+//
+// See 'man 2 bpf' for details.
+template <typename KeyT, typename ValueT>
+inline error::StatusOr<posix::UniqueFileDescriptor> MapCreate(
+    const uint32_t map_type, const size_t max_entries) {
```
Comment (epeer-juniper) [inline lib/bpf/map_create.h]: Same question with respect to map_type. Do we want to have stronger typing at this level already?

Comment (epeer-juniper) [reply lib/bpf/map_create.h]: In terms of consistency, we are already exposing a more specific return type in this instance (when compared to the Bpf() call)

File: lib/bpf/bpf.h (hunk 1 above)
Comment (epeer-juniper) [reply lib/bpf/bpf.h]: To answer my own question a bit, I guess Bpf() is multiplexed, which would force a bunch of logic into the wrapper if you wanted to be specific on the return type. That said, perhaps Bpf() should be more hidden?

File: lib/bpf/prog_load.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace bpf {
+
+// BPF_PROG_LOAD wrapper.
+//
+// TODO: Incomplete.
+//
+// See 'man 2 bpf' for details.
+inline error::StatusOr<posix::UniqueFileDescriptor> ProgLoad(
+    const uint32_t program_type, std::string_view instructions,
```
Comment (epeer-juniper) [inline lib/bpf/prog_load.h]: Tighter types for program_type? And, program_flags?

File: lib/bpf/bpf.h (hunk 1 above)
Comment (jscherpelz-juniper) [reply lib/bpf/bpf.h]: You make a great point, and I'm trying to establish a standard of sorts. Here's what I'm proposing for now:

* syscalls (both with and without C standard library wrappers) must be wrapped such that they return error::Status or error::StatusOr. 
* The wrapper should be spelled the same but with Google C++ style capitalization and all underscores removed. When a prefix of the name is implied by the containing namespace it should be omitted (for example BPF_*).
* Inputs should remain unchanged with a few exceptions:
    - use std::string_view for input that is interpreted as a const array of bytes.
    - use std::string for anything that the syscall interprets as a null terminated string (std::string::c_str() makes this easy but std::string_view has no such equivalent).
    - File descriptors must be wrapped. The goal is to call GetValue() on a wrapped type as close to the point of consumption as possible.
    - output arrays of bytes must be wrapped in a std::span equivalent.

I consider any function that takes a pointer and a length/size to be an antipattern since it is so easy to make a mistake. This is the motivation for most of the above rules that alter function signatures.

Additional wrappers can then be provided to make the syscalls more useful and even better typed. The general idea is to make the lowest level syscall wrappers look like the original and do minimal work. As more layers are added eventually a user should have no idea what syscall is being invoked.

With respect to 'int cmd' and and 'uint32_t map_type', these represent a difficult situation. If we make an enum then it must be kept up to date (automation can help but there can still be version conflicts), and it will confuse users who are familiar with the existing constant names. It makes googling for the meaning of a constant harder because of the added indirection. It is also possible to wrap cmd and map_type in base::OpaqueValue, but this also complicates the interface without providing much value (the values are wrapped just before use so there is little chance of misuse).

The rule of thumb I'm using is that if a type is copied directly from the output of a syscall to the input of another syscall without interpretation then it should be wrapped (see FileDescriptor and IfIndex). However a constant that is used as the input to a single function is worth keeping in sync with the system/kernel headers.

File: lib/posix/syscall.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+template <typename... ArgsT>
+inline error::StatusOr<int> Syscall(const int number, ArgsT&&... args) {
+  const auto rv = ::syscall(number, std::forward<ArgsT>(args)...);
+  RETURN_IF_ERROR(OkStatusOrCaptureErrnoIf(-1 == rv, "syscall() failed"));
+  return rv;
+}
+
+}  // namespace posix
+
+#endif  // LIB_POSIX_SYSCALL_H_
```
Comment (epeer-juniper) [inline lib/posix/syscall.h]: This should probably be a separate commit.

File: lib/bpf/map_create.h (hunk 2 above)
Comment (jscherpelz-juniper) [reply lib/bpf/map_create.h]: See the comment on 'int cmd' and let me know what you think

File: lib/bpf/prog_load.h (hunk 3 above)
Comment (jscherpelz-juniper) [reply lib/bpf/prog_load.h]: See the comment on 'int cmd' and let me know what you think

File: lib/posix/unique_file_descriptor.h [hunk 5]
```diff
@@ -41,6 +41,14 @@ struct UniqueFileDescriptorDtor {
...
     base::UniqueValue<FileDescriptor, impl::UniqueFileDescriptorDtor>;
 
+// Idiom to wrap a raw integer fd in a UniqueFileDescriptor.
+// Takes ownership of 'raw_fd'.
+inline UniqueFileDescriptor MakeUniqueFileDescriptor(const int raw_fd) {
+  const FileDescriptor fd(raw_fd);
+  INVARIANT_T(IsWellFormed(fd));
+  return UniqueFileDescriptor(fd);
+}
+
```
Comment (epeer-juniper) [inline lib/posix/unique_file_descriptor.h]: This too.

File: lib/bpf/bpf.h (hunk 1 above)
Comment (epeer-juniper) [reply lib/bpf/bpf.h]: Perhaps the choice of namespace should make it more clear that these functions are not intended to higher level users? Namespace bpf in a project where BPF is a top level concept might have people looking here when they should should instead be looking for higher level APIs.

Comment (jscherpelz-juniper) [reply lib/bpf/bpf.h]: I've been playing with a ::bpf::raw and ::posix::raw namespace to use for first level syscall wrappers. Any thoughts?

File: lib/bpf/prog_load.h (hunk 3 above)
Comment (epeer-juniper) [reply lib/bpf/prog_load.h]: This one is more grey, because ProgLoad() is no longer a 1:1 mapping to a syscall. What is the ProgLoad() called that I should use as an end user of the library that just wants to load a piece of eBPF code? What encapsulates (in the hide sense) this?

File: lib/bpf/bpf.h (hunk 1 above)
Comment (epeer-juniper) [reply lib/bpf/bpf.h]: I like.

Comment (epeer-juniper) [reply lib/bpf/bpf.h]: Or the other way around? raw::bpf and raw::posix?

Comment (epeer-juniper) [reply lib/bpf/bpf.h]: Another idea, sys::wapper::posix and sys::wrapper::bpf?

File: lib/bpf/prog_load.h (hunk 3 above)
Comment (jscherpelz-juniper) [reply lib/bpf/prog_load.h]: The hope is that the user of a library is still a few levels above this. This is the part of the design that isn't clear to me yet. My approach for now is to wrap kernel functionality with names that mirror the syscall/command constants until it is easy and obvious what the final function should be called. At a high enough level we'll start to have more traditional OOP classes and users won't need to worry about things like file descriptors and syscalls.

I would like to consider the function names so far to be uncontroversial in the sense that they are mechanical transformations of names from kernel headers. If you have other ideas I'd love to hear them. I think namespaces are often a fantastic tool to solve problems like this.

File: lib/bpf/bpf.h (hunk 1 above)
Comment (jscherpelz-juniper) [reply lib/bpf/bpf.h]: I'd love to use the sys namespace, but that is already part of the standard library and I'd rather not pollute it.

I do like changing the order to put ::raw first. This also makes the standard for stripping namespace prefixes off of function names more sensible.

Comment (anandrao79) [inline lib/bpf/bpf.h]: With this, are we moving away from using libbpf for userspace?

File: lib/bpf/prog_load.h [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  attr.prog_type = program_type;
+  attr.insn_cnt = instructions.size() / kInstructionSize;
+  attr.insns = reinterpret_cast<uintptr_t>(instructions.data());
+  attr.license = reinterpret_cast<uintptr_t>(license.c_str());
+  // TODO: Log buffer? Need std::span<char> equivalent.
+  // TODO: Kernel version? Only needed for perf programs?.
+  attr.prog_flags = program_flags;
+  strncpy(attr.prog_name, name.c_str(), sizeof(attr.prog_name));
+  attr.prog_ifindex = GetValue(ifindex);
+  ASSIGN_OR_RETURN(const auto fd, Bpf(BPF_PROG_LOAD, &attr));
```
Comment (anandrao79) [inline lib/bpf/prog_load.h]: With this, we don't need the libbpf way of loading the xdp programs? Isn't it?

File: lib/bpf/bpf.h (hunk 1 above)
Comment (jscherpelz-juniper) [reply lib/bpf/bpf.h]: I think it is too early to say. All I'm doing is experimenting with the kernel interface to better understand what is going on and checking in the utilities I make along the way.

I suspect that we may want to manipulate BPF maps directly, but it is not certain.

File: lib/bpf/prog_load.h (hunk 6 above)
Comment (jscherpelz-juniper) [reply lib/bpf/prog_load.h]: As mentioned above for maps, the goal isn't necessarily to replace libbpf, just to understand the kernel interface better. Certainly ProgLoad() is missing a ton of features compared to libbpf.

In the case of ProgLoad() I suspect that once we start creating maps of type BPF_MAP_TYPE_PROG_ARRAY we may want to load programs more directly?

File: lib/bpf/prog_load.h (hunk 3 above)
Comment (epeer-juniper) [reply lib/bpf/prog_load.h]: Yes, I think this is addressed by making it clear that the namespace is not for end user consumption (raw::).

**Comment by jscherpelz-juniper** [comment]
Closing PR. This content will return in a series of more polished PRs
